zarr/storage.py

```diff
@@ -1877,6 +1878,197 @@ def __delitem__(self, key):
             self._invalidate_value(key)
 
 
+class SQLiteStore(MutableMapping):
+    """Storage class using SQLite.
+
+    Parameters
+    ----------
+    path : string
+        Location of database file.
+    **kwargs
+        Keyword arguments passed through to the `sqlite3.connect` function.
+
+    Examples
+    --------
+    Store a single array::
+
+        >>> import zarr
+        >>> store = zarr.SQLiteStore('data/array.sqldb')
+        >>> z = zarr.zeros((10, 10), chunks=(5, 5), store=store, overwrite=True)
+        >>> z[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+
+    Store a group::
+
+        >>> store = zarr.SQLiteStore('data/group.sqldb')
+        >>> root = zarr.group(store=store, overwrite=True)
+        >>> foo = root.create_group('foo')
+        >>> bar = foo.zeros('bar', shape=(10, 10), chunks=(5, 5))
+        >>> bar[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+    """
+
+    def __init__(self, path, **kwargs):
+        import sqlite3
+
+        kwargs.setdefault('timeout', 5.0)
+        kwargs.setdefault('cached_statements', 100)
+
+        # normalize path
+        if path != ':memory:':
+            path = os.path.abspath(path)
+
+        # store properties
+        self.path = path
+        self.kwargs = kwargs
+
+        # open database
+        self.db = sqlite3.connect(
+            self.path,
+            detect_types=0,
+            isolation_level=None,
+            check_same_thread=False,
```

Comment:
Didn't we decide it should be `check_same_thread=True`?

Reply:
Hm, this is messy. I think there may be two separate issues here. 

The first issue is that older versions of sqlite3 cannot share connections between threads.

However, there is a second issue (which I'm more concerned about), which is that even if you have a newer version of sqlite3, I'm still not sure it's a good idea to allow multiple threads to use the same connection if there is any chance of multiple threads attempting to write at the same time, even if threads are writing different chunks of an array.

This sentence from the [Python 3 docs](https://docs.python.org/3.7/library/sqlite3.html#sqlite3.connect) suggests concurrent writes are not good:

> When using multiple threads with the same connection writing operations should be serialized by the user to avoid data corruption.

Also [this page on the sqlite docs](https://www.sqlite.org/threadsafe.html) describes how there are three different threading modes, "single-threaded", "multi-threaded" and "serialized". It's not immediately clear which of these three modes is the default when using sqlite3 via Python standard library (would be good to know that).

So I think the situation is that:

* If a user wants to **read** data from a zarr store using multiple threads, that should be fine, as long as they don't have an ancient version of sqlite. I.e., setting `check_same_thread=False` is fine.
* If a user wants to **write** data into a zarr store using multiple threads, that should be fine as long as sqlite3 is running in "serialized" mode (in which case setting `check_same_thread=False` would also be fine). However, if sqlite3 is not running in "serialized" mode, then we might need to think about what to do.

Reply:
On the first point, have pushed a small change to check that the SQLite version is new enough before setting `check_same_thread=False`. Hopefully that sufficiently addresses it.

On the second point, it looks like [the SQLite docs]( https://www.sqlite.org/threadsafe.html#overview ) says the following.

> The default mode is serialized.

The threading mode could be changed in custom SQLite builds. By default it builds with serialized mode. Seems that's how we are doing it in conda-forge. IDK how Python.org does it, but would guess it is the same.

Also the threading mode could be selected when initializing the SQLite application and/or for each connection opened. In either case the default for both is serialized mode. CPython uses the default in both cases.

Reply:
Yeah that does suggest "serialized" mode will be used when using sqlite3 via Python standard library, but if so then why do they say in the Python 3 docs, "When using multiple threads with the same connection writing operations should be serialized by the user to avoid data corruption"? If "serialized" mode really does what is says then you ought to be able to hammer an sqlite3 database from multiple writer threads and serialization will be handled internally within the sqlite3 library. I'm a bit confused.

In any case, I'm happy to go with `check_same_thread=False` and let the rest come out in the wash.

Reply:
Agree the documentation is a little confusing on this matter. It also seems a bit dated and a bit short on thread safety, which doesn't help.

In the interest of being cautious, we could try creating a lock and holding it while performing any mutation on the database. FWICT it doesn't seem to be needed, but it is easy to do and a small price to pay if it provides more confidence about the store's behavior. As a footnote Python locks can't be pickled, but it doesn't sound like we would need this one to be.

Reply:
Good idea, why don't we add a lock, as you say it should be simple and shouldn't have any downside, and will give us some confidence that the store is safe to use with multiple writer threads.

Reply:
...and yes I don't think it's a problem that the lock cannot be pickled, it should be fine to instantiate a new `Lock` when the store's ``__init__()`` is called.

Reply:
Cool, have added a lock. Please let me know if that works.